src/container/srv_container.c

```diff
@@ -960,6 +971,308 @@ cont_destroy(struct rdb_tx *tx, struct ds_pool_hdl *pool_hdl,
+		D_FREE(ec_agg->ea_server_ephs);
+	D_FREE(ec_agg);
+}
+
+/**
+ * Update the epoch (by rank) on the leader of container service, which
+ * will be called by IV update on the leader.
+ */
+int
+ds_cont_leader_update_agg_eph(uuid_t pool_uuid, uuid_t cont_uuid,
+			      d_rank_t rank, daos_epoch_t eph)
+{
+	struct cont_svc		*svc;
+	struct cont_ec_agg	*ec_agg;
+	int			rc;
+	bool			retried = false;
+	int			i;
+
+	rc = cont_svc_lookup_leader(pool_uuid, 0 /* id */, &svc,
+				    NULL /* hint */);
+	if (rc != 0)
+		return rc;
+
+retry:
+	ec_agg = cont_ec_agg_lookup(svc, cont_uuid);
+	if (ec_agg == NULL) {
+		rc = cont_ec_agg_alloc(svc, cont_uuid, &ec_agg);
+		if (rc)
+			D_GOTO(out_put, rc);
+	}
+
+	for (i = 0; i < ec_agg->ea_servers_num; i++) {
+		if (ec_agg->ea_server_ephs[i].rank == rank) {
+			if (ec_agg->ea_server_ephs[i].eph < eph)
+				ec_agg->ea_server_ephs[i].eph = eph;
+			break;
+		}
+	}
+
+	if (i == ec_agg->ea_servers_num) {
+		if (!retried) {
+			D_DEBUG(DB_MD, "rank %u eph "DF_U64" retry for"
+				DF_CONT"\n", rank, eph,
+				DP_CONT(pool_uuid, cont_uuid));
+			retried = true;
+			cont_ec_agg_destroy(ec_agg);
+			goto retry;
+		} else {
+			D_WARN("rank %u eph "DF_U64" does not exist for "
+			       DF_CONT"\n", rank, eph,
+			       DP_CONT(pool_uuid, cont_uuid));
+		}
+	} else {
+		D_DEBUG(DB_MD, DF_CONT" update eph rank %u eph "DF_U64"\n",
+			DP_CONT(pool_uuid, cont_uuid), rank, eph);
+	}
+
+out_put:
+	cont_svc_put_leader(svc);
+	return 0;
+}
+
+struct refresh_vos_agg_eph_arg {
+	uuid_t	pool_uuid;
+	uuid_t  cont_uuid;
+	daos_epoch_t min_eph;
+};
+
+int
+cont_refresh_vos_agg_eph_one(void *data)
+{
+	struct refresh_vos_agg_eph_arg *arg = data;
+	struct ds_cont_child	*cont_child;
+	int			rc;
+
+	rc = ds_cont_child_lookup(arg->pool_uuid, arg->cont_uuid, &cont_child);
+	if (rc)
+		return rc;
+
+	D_DEBUG(DB_MD, DF_CONT": update aggregation max eph "DF_U64"\n",
+		DP_CONT(arg->pool_uuid, arg->cont_uuid), arg->min_eph);
+	cont_child->sc_ec_agg_eph_boundry = arg->min_eph;
+	ds_cont_child_put(cont_child);
+	return rc;
+}
+
+int
+ds_cont_tgt_refresh_agg_eph(uuid_t pool_uuid, uuid_t cont_uuid,
+			    daos_epoch_t eph)
+{
+	struct refresh_vos_agg_eph_arg	arg;
+	int				rc;
+
+	uuid_copy(arg.pool_uuid, pool_uuid);
+	uuid_copy(arg.cont_uuid, cont_uuid);
+	arg.min_eph = eph;
+
+	rc = dss_task_collective(cont_refresh_vos_agg_eph_one, &arg, 0,
+				 DSS_ULT_IO);
+	return rc;
+}
+
+#define EC_AGG_EPH_INTV	 (10ULL * NSEC_PER_SEC)	/* seconds interval to check*/
+static void
+cont_agg_eph_leader_ult(void *arg)
+{
+	struct dss_sleep_ult	*sleep_ult;
+	struct cont_svc		*svc = arg;
+	int			rc = 0;
+
+	sleep_ult = dss_sleep_ult_create();
+	if (sleep_ult == NULL) {
+		D_ERROR(DF_UUID " sleep ult creation failed.\n",
+			DP_UUID(svc->cs_pool->sp_uuid));
+		return;
+	}
+
+	while (1) {
+		struct cont_ec_agg	*ec_agg;
+		d_rank_list_t		fail_rank_list = { 0 };
+
+		rc = map_ranks_init(svc->cs_pool->sp_map, MAP_RANKS_DOWN,
+				    &fail_rank_list);
```

Comment:
Looks this function is to check node state. If all targets of a node are DOWN, will we change the node state to DOWN or keep it as UP?